pysb/core.py

```diff
@@ -1302,25 +1302,31 @@ def value(self):
 
     @value.setter
     def value(self, new_value):
-        if self.is_integer:
-            if not float(new_value).is_integer():
-                raise ValueError('Cannot assign an non-integer value to a '
-                                 'parameter assumed to be an integer')
-        if self.is_nonnegative:
-            if float(new_value) < 0:
-                raise ValueError('Cannot assign a negative value to a '
-                                 'parameter assumed to be nonnegative')
+        _check_value_consistency(new_value, self.is_integer,
+                                 self.is_nonnegative)
 
         self._value = float(new_value)
     
     def get_value(self):
         return self.value
 
     def __repr__(self):
-        return  '%s(%s, %s)' % (self.__class__.__name__, repr(self.name), repr(self.value))
+        return '%s(%s, %s)' % (self.__class__.__name__, repr(self.name), repr(self.value))
 
     def __str__(self):
-        return  repr(self)
+        return repr(self)
+
+
+def _check_value_consistency(value, is_integer, is_nonnegative):
```

Comment:
I like the idea of refactoring the checking logic out of the property setter so the simulator can call it. What about making this a method on `Parameter`, maybe `def check_value(value)`? Then `Parameter.__init__` could call `self.check_value(value)` and the simulator code could call `param.check_value(value)`.

The exception message should maybe include the parameter name and the value, to help in locating the actual problem when the problem arises in the simulator calls (where the parameter name and value are often coming from a separate data source and so the call stack won't help identify the problem). Or you can leave these messages alone, then catch these exceptions in the simulator `_check_parameter_value` helper and re-raise them with augmented messages. Making these a custom subclass of ValueError would help with that.

Reply:
Good point, initially thought about adding to `__new__` which didn't work as assumptions were not set. Checking in `__ini__` does indeed work with a class method. Also realized that I already added testcases that checked for bad value in `__init__`, which already passed. `__init__` calls the setter so no additional checks are necessary there ;)